manifest: Give the option to get BabbleSim via the manifest

aescolar opened the pull request:
To improve users experience, give the option to users to fetch the required BabbleSim components using the west manifest.
Until now, users would need to either use the Zephyr docker image, or fetch bsim on their own.
For many years there has been a request to automate this process. This is the first step.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/55698

Note: A set of follow ups could/would:
* For developers ease, have cmake find the simulator thru west if the environment variables that tell where
the simulator is are not set.  https://github.com/aescolar/zephyr/commit/9e331be4c79a1cdab7e0ad0a3cfe2cb36017e575 
* Add a cmake check to check the simulator is the necessary version, and otherwise instruct users to west update and rebuild it: https://github.com/aescolar/zephyr/commit/d9a49dced8f6b9c7b6dfbd4f017bca7c2e1d2155
* Have CI use the version of the simulator in the manifest

aescolar committed:
manifest: Get BabbleSim via the manifest

To facilitate users experience, fetch the required
BabbleSim components using the west manifest.
Until now, users would need to either use the Zephyr
docker image, or fetch bsim on their own.
For many years there has been a request to automate this
process. This is the first step.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

stephanosio commented on west.yml:
```diff
@@ -262,6 +264,13 @@ manifest:
     - name: thrift
       path: modules/lib/thrift
       revision: 10023645a0e6cb7ce23fcd7fd3dbac9f18df6234
+    - name: bsim
+      revision: da8ca8270cc14010ff1deda851112b7a04ec562e
+      # This URL would not be here, instead this manifest repo would be hosted
+      # in the zephyr-rtos org. This is here just to enable a functional demo.
+      url: https://github.com/BabbleSim/bsim_west.git
+      import:
+        path-prefix: tools
```
```
# Please add items below based on alphabetical order
```

jori-nordic commented:
Ran the gatt settings test successfully with the bsim fetched using this PR.

stephanosio committed:
ci: doc-build: Pull 'bsim' west project

This commit updates the documentation build workflow to pull in the
'bsim' west project (with its submanifest) so that the build process
does not fail due to a manifest import failure.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>
Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

stephanosio commented:
@aescolar Regarding the following documentation build failure, you can cherry-pick https://github.com/zephyrproject-rtos/zephyr/pull/56160/commits/6c0b69559bb8edab3be31b1f46f861e23b04ebc8 to fix the issue.

```
make: Entering directory '/home/runner/work/zephyr/zephyr/doc'
make html DT_TURBO_MODE=1
make[1]: Entering directory '/home/runner/work/zephyr/zephyr/doc'
cmake \
	-GNinja \
	-B_build \
	-S. \
	-DDOC_TAG=development \
	-DSPHINXOPTS="-q -W -t publish" \
	-DLATEXMKOPTS="-halt-on-error -no-shell-escape" \
	-DDT_TURBO_MODE=1
Loading Zephyr module(s) (Zephyr repository): doc
-- Found Python3: /usr/bin/python3.10 (found suitable exact version "3.10.6") found components: Interpreter 
-- Cache files will be written to: /home/runner/.cache/zephyr
-- Found west (found suitable version "1.0.0", minimum required is "1.0.0")
CMake Error at /home/runner/work/zephyr/zephyr/cmake/modules/zephyr_module.cmake:77 (message):
  Traceback (most recent call last):

    File "/home/runner/work/zephyr/zephyr/scripts/zephyr_module.py", line 718, in <module>
      main()
    File "/home/runner/work/zephyr/zephyr/scripts/zephyr_module.py", line 663, in main
      west_projs = west_projects()
    File "/home/runner/work/zephyr/zephyr/scripts/zephyr_module.py", line 531, in west_projects
      manifest = Manifest.from_file()
    File "/home/runner/.local/lib/python3.10/site-packages/west/manifest.py", line 1170, in from_file
      return Manifest(topdir=topdir, config=config,
    File "/home/runner/.local/lib/python3.10/site-packages/west/manifest.py", line 1396, in __init__
      self._load_validated()
    File "/home/runner/.local/lib/python3.10/site-packages/west/manifest.py", line 1791, in _load_validated
      self._load_projects(manifest_data, url_bases, defaults)
    File "/home/runner/.local/lib/python3.10/site-packages/west/manifest.py", line [21](https://github.com/zephyrproject-rtos/zephyr/actions/runs/4502917193/jobs/7925415081?pr=55696#step:7:22)14, in _load_projects
      self._import_from_project(project, imp)
    File "/home/runner/.local/lib/python3.10/site-packages/west/manifest.py", line 2300, in _import_from_project
      self._import_map_from_project(project, imp)
    File "/home/runner/.local/lib/python3.10/site-packages/west/manifest.py", line 2327, in _import_map_from_project
      imported = self._import_content_from_project(project, imap.file)
    File "/home/runner/.local/lib/python3.10/site-packages/west/manifest.py", line 2395, in _import_content_from_project
      content = self._ctx.project_importer(project, path)
    File "/home/runner/.local/lib/python3.10/site-packages/west/manifest.py", line 164, in _default_importer
      raise ManifestImportFailed(project, file)

  west.manifest.ManifestImportFailed: ManifestImportFailed: project <Project
  bsim ('/home/runner/work/zephyr/tools/bsim') at
  bc[22](https://github.com/zephyrproject-rtos/zephyr/actions/runs/4502917193/jobs/7925415081?pr=55696#step:7:23)891208d453a8cb851e4ecdc8f165f3a33890> value west.yml

Call Stack (most recent call first):
  /home/runner/work/zephyr/zephyr/cmake/modules/doc.cmake:24 (include)
  /home/runner/work/zephyr/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:75 (include)
  /home/runner/work/zephyr/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:117 (include_boilerplate)
  CMakeLists.txt:7 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/runner/work/zephyr/zephyr/doc/_build/CMakeFiles/CMakeOutput.log".
make[1]: *** [Makefile:[23](https://github.com/zephyrproject-rtos/zephyr/actions/runs/4502917193/jobs/7925415081?pr=55696#step:7:24): configure] Error 1
make[1]: Leaving directory '/home/runner/work/zephyr/zephyr/doc'
make: Leaving directory '/home/runner/work/zephyr/zephyr/doc'
make: *** [Makefile:17: html-fast] Error 2
```

carlescufi commented:
I am missing the place where you execute `west config manifest.group-filter +babblesim`

carlescufi commented on .github/workflows/doc-build.yml:
```diff
@@ -69,6 +69,7 @@ jobs:
     - name: west setup
       run: |
         west init -l .
+        west update bsim
```
I don't get how this is fixing anything. By default the `babblesim` group is disabled, so this will be a nop?

aescolar replied:
this is just fetching the bsim manifest repo (which has submanifest), that one is not under the group

stephanosio replied:
No problem with that as long as that works; though, why the `west init -l .` above does not fetch the bsim manifest repo is still a mystery to me.

carlescufi commented:
> I am missing the place where you execute `west config manifest.group-filter +babblesim`

Sorry, I failed to remember that we are not pulling BabbleSim from the repo ourselves

aescolar commented:
I just realized the pdf docs build needed the same fix as the html docs build

carlescufi reviewed (dismissed):
This looks good now.

gmarull commented on .github/workflows/doc-build.yml:
```diff
@@ -69,6 +69,7 @@ jobs:
     - name: west setup
       run: |
         west init -l .
+        west update bsim
```
I'm not sure I follow here, how is bsim related to docs?

carlescufi replied:
It's not, but the docs are using `west` and west needs to resolve the manifest. Because we now have an `import` statement in the main manifest, we need the `west update bsim` to pull the manifest that gets `import`ed

carlescufi replied:
The key here is that, for the first time, we have an `import` statement in the default upstream manifest.

aescolar replied:
It's because of this error:
https://github.com/zephyrproject-rtos/zephyr/pull/55696#issuecomment-1481530833
That is elaborated a bit more here:
https://github.com/zephyrproject-rtos/west/issues/644

west init does not fetch the project which has the sub-manifest, so following west commands that look thru the manifest would error out as they will not find that imported manifest). The solution is to do either a full west update after west init, or to fetch just the missing project which has that sub-manifest

gmarull replied:
does that mean this will impact all users then? too me it looks like a workaround due to a bug, am I wrong? If that's the case please add a `FIXME`.

carlescufi replied:
I don't think this is a bug, no. `west init` works and `west update` works fine. What has changed is that you cannot execute things like `west list` before `west update`, but that has never been something we've publicized.

carlescufi replied:
The difference is that now you if you do something like `west list` before you run `west update` you will get:

```
❯ west list
FATAL ERROR: failed manifest import in bsim (tools/bsim)
  Failed importing "west.yml" from revision "bc22891208d453a8cb851e4ecdc8f165f3a33890"
  Hint: for this to work:
          - bsim must be cloned
          - its manifest-rev ref must point to a commit with the import data
        To fix, run:
          west update
```
or an exception thrown if you do it via the west API (like you saw in the doc-build workflow). But this is not a bug, this is the correct west behavior as it stands now. What @aescolar proposes in https://github.com/zephyrproject-rtos/west/issues/644 is an enhancement.

gmarull replied:
> I don't think this is a bug, no. `west init` works and `west update` works fine. What has changed is that you cannot execute things like `west list` before `west update`, but that has never been something we've publicized.

but we're just running `west init --local .`, how is west list involved here?

carlescufi replied:
> @gmarull the issue was triggered in the documentation build cmake files which use zephyr/cmake/modules/zephyr_module.cmake

Correct, the script `zephyr_module.cmake` invokes `zephyr_module.py` which in turn invokes `west`. This is necessary to generate all the `Kconfig.modules` stuff that Kconfig needs to parse the Kconfig symbols, among other things.

nashif commented on west.yml:
```diff
@@ -22,9 +22,16 @@ manifest:
   # Please add items below based on alphabetical order
   projects:
+    - name: bsim
+      repo-path: babblesim-manifest
```
needs further discussion, this manifest repo includes repos not hosted by the zephyr project, something that is not covered by the current guidelines and rules for includng repos in the manifest.

alwa-nordic replied:
Can we set up a Zephyrproject mirror for the repo? Who do we need to talk to? @carlescufi, do you know?

aescolar commented:
Added DNM following @nashif request for further discussion

tejlmand commented on boards/posix/nrf52_bsim/find_babblesim.cmake:
```diff
@@ -0,0 +1,45 @@
+# Copyright (c) 2023 Nordic Semiconductor ASA
```
Good direction, but I would like this to become a CMake module.

That means relocate this file to https://github.com/zephyrproject-rtos/zephyr/tree/main/cmake/modules and name it:
`Findbabblesim.cmake`.

That way users can lookup babblesim using:
```
find_package(babblesim)
```

let's take other details offline.

aescolar replied:
The cmake changes will be taken in another PR.

tejlmand reviewed (dismissed):
Good direction.

Let's take this one step further 😃

marc-hb commented:
This breaks `west update_only_what_I_need` for all users:
- #57198
